File: frontend/src/engine/capabilities/lookupAll.ts
/**
 * lookupAll – Find ALL matching rows between two ranges, not just the first.
 *
 * For each lookup value, finds every matching row in the source range and
 * collects values from the specified return column. Results are joined with
 * a delimiter and written to the output range.
 */

import { CapabilityMeta, LookupAllParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "lookupAll",
  description: "Find all matching rows between two ranges and return joined values",
  mutates: true,
  affectsFormatting: false,
};

async function handler(
  context: Excel.RequestContext,
  params: LookupAllParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  const { lookupRange, sourceRange, returnColumn, outputRange, delimiter = ", ", matchType = "exact" } = params;

  if (options.dryRun) {
    return {
      stepId: "",
      status: "success",
      message: `Would look up all matches from ${lookupRange} in ${sourceRange}, return column ${returnColumn} to ${outputRange}`,
    };
  }

  options.onProgress?.("Reading lookup and source ranges...");

  const lookupRng = resolveRange(context, lookupRange).getUsedRange(false);
  const sourceRng = resolveRange(context, sourceRange).getUsedRange(false);
  lookupRng.load("values");
  sourceRng.load("values");
  await context.sync();

  const lookupVals = (lookupRng.values ?? []) as (string | number | boolean | null)[][];
  const sourceVals = (sourceRng.values ?? []) as (string | number | boolean | null)[][];

  if (!lookupVals.length || !sourceVals.length) {
    return { stepId: "", status: "success", message: "No data to look up." };
  }

  const retColIdx = returnColumn - 1; // convert 1-based to 0-based

  options.onProgress?.(`Looking up ${lookupVals.length} values against ${sourceVals.length} source rows...`);

  // Build an index: source key (first column) → array of return column values
  const index = new Map<string, string[]>();
  for (const row of sourceVals) {
    const key = String(row[0] ?? "").trim().toLowerCase();
    if (!key) continue;
    const retVal = String(row[retColIdx] ?? "");
    if (!index.has(key)) index.set(key, []);
    index.get(key)!.push(retVal);
  }

  const results: (string | null)[][] = [];
  let matchedLookups = 0;
  let totalMatches = 0;

  for (const row of lookupVals) {
    const lookupStr = String(row[0] ?? "").trim().toLowerCase();
    if (!lookupStr) {
      results.push([null]);
      continue;
    }

    let matched: string[] = [];

    if (matchType === "exact") {
      matched = index.get(lookupStr) ?? [];
    } else {
      // contains: find all source keys that contain or are contained by the lookup value
      for (const [key, vals] of index) {
        if (key.includes(lookupStr) || lookupStr.includes(key)) {
          matched.push(...vals);
        }
      }
    }

    if (matched.length > 0) {
      results.push([matched.join(delimiter)]);
      matchedLookups++;
      totalMatches += matched.length;
    } else {
      results.push([null]);
    }
  }

  // Write results to output range
  const outRng = resolveRange(context, outputRange);
  outRng.getResizedRange(results.length - 1, 0).values = results as any;
  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Found matches for ${matchedLookups}/${lookupVals.length} lookup values (${totalMatches} total matches)`,
    outputs: { outputRange },
  };
}

registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/lookup_all.py
"""lookupAll — return ALL matching rows between two ranges joined by a delimiter.

Port of `frontend/src/engine/capabilities/lookupAll.ts`. For each lookup value,
finds every matching source row (by first column) and collects values from the
specified 1-based return column. Results are joined with a delimiter and
written to the output range.
"""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


def _as_2d(raw: Any, shape: tuple[int, int]) -> list[list[Any]]:
    if raw is None:
        return []
    rows_, cols_ = shape
    if not isinstance(raw, list):
        return [[raw]]
    if raw and not isinstance(raw[0], list):
        if rows_ == 1:
            return [list(raw)]
        return [[v] for v in raw]
    return [list(r) for r in raw]


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    lookup_range = params.get("lookupRange")
    source_range = params.get("sourceRange")
    return_column = params.get("returnColumn")
    output_range = params.get("outputRange")
    delimiter = params.get("delimiter", ", ")
    match_type = (params.get("matchType") or "exact").lower()

    if not lookup_range or not source_range or return_column is None or not output_range:
        return {
            "status": "error",
            "message": (
                "lookupAll requires 'lookupRange', 'sourceRange', 'returnColumn', "
                "and 'outputRange'."
            ),
        }

    if ctx.dry_run:
        return {
            "status": "preview",
            "message": (
                f"Would look up all matches from {lookup_range} in {source_range}, "
                f"return column {return_column} to {output_range}."
            ),
        }

    try:
        lookup_rng = resolve_range(ctx.workbook_handle, lookup_range)
        source_rng = resolve_range(ctx.workbook_handle, source_range)

        lookup_vals = _as_2d(lookup_rng.value, lookup_rng.shape)
        source_vals = _as_2d(source_rng.value, source_rng.shape)

        if not lookup_vals or not source_vals:
            return {"status": "success", "message": "No data to look up.", "outputs": {}}

        ret_col_idx = int(return_column) - 1

        # Build an index: source key (first column) → list of return-column values.
        index: dict[str, list[str]] = {}
        for row in source_vals:
            if not row:
                continue
            key = str(row[0] if row[0] is not None else "").strip().lower()
            if not key:
                continue
            ret_val_raw = row[ret_col_idx] if 0 <= ret_col_idx < len(row) else None
            ret_val = str(ret_val_raw if ret_val_raw is not None else "")
            index.setdefault(key, []).append(ret_val)

        results: list[list[Any]] = []
        matched_lookups = 0
        total_matches = 0

        for row in lookup_vals:
            lookup_str = str(row[0] if row and row[0] is not None else "").strip().lower()
            if not lookup_str:
                results.append([None])
                continue

            matched: list[str] = []
            if match_type == "exact":
                matched = list(index.get(lookup_str, []))
            else:
                # "contains": either direction — lookup contains key OR key contains lookup.
                for key, vals in index.items():
                    if lookup_str in key or key in lookup_str:
                        matched.extend(vals)

            if matched:
                results.append([delimiter.join(matched)])
                matched_lookups += 1
                total_matches += len(matched)
            else:
                results.append([None])

        out_rng = resolve_range(ctx.workbook_handle, output_range)
        out_rng.resize(len(results), 1).value = results
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"lookupAll failed: {exc}", "error": str(exc)}

    return {
        "status": "success",
        "message": (
            f"Found matches for {matched_lookups}/{len(lookup_vals)} lookup values "
            f"({total_matches} total matches)."
        ),
        "outputs": {"outputRange": output_range},
    }


registry.register("lookupAll", handler, mutates=True)
